### PR TITLE
Add a downgrade utility script to automate the procedure

### LINKED ISSUE #1403: Create a script to downgrade metalk8s cluster
**Component**:

'salt'

**Why this is needed**:

Currently the process to upgrade is really not user friendly

**What should be done**:

Create a script to simplify all this process and update the doc accordingly 
https://github.com/scality/metalk8s/blob/development/2.0/docs/operation/downgrade.rst

**Procedure**:

Downgrade:

`<current_cluster_version>` is the highest version available in the cluster

- Connect to the salt-master container
```
crictl exec -it $(crictl ps -q --label io.kubernetes.container.name=salt-master) bash
```

- To check what will be done (replace `<destination_version>`)
```
salt-run state.orchestrate metalk8s.orchestrate.downgrade.precheck saltenv=metalk8s-<current_cluster_version> pillar="{'orchestrate': {'dest_version': '<destination_version>'}}"
```

- Launch the downgrade
```
salt-run state.orchestrate metalk8s.orchestrate.downgrade saltenv=metalk8s-<current_cluster_version> pillar="{'orchestrate': {'dest_version': '<destination_version>'}}"
```

(Leave the salt-master container)
- Downgrade the bootstrap node (replace `<destination_version>`)
```
# Downgrade bootstrap stuff
salt-call --local state.sls metalk8s.roles.bootstrap saltenv=metalk8s-<destination_version> pillar="{'metalk8s': {'endpoints': $(salt-call --out txt pillar.get metalk8s:endpoints | cut -c 8-)}}"

# Connect to salt-master
crictl exec -it $(crictl ps -q --label io.kubernetes.container.name=salt-master) bash

# Set the version in kubernetes nodes object
salt-run salt.cmd state.single metalk8s_kubernetes.node_label_present metalk8s.scality.com/version node=<node_name> value=<dest_version> kubeconfig=/etc/kubernetes/admin.conf
# Run downgrade on bootstrap node 
salt-run state.sls metalk8s.orchestrate.deploy_node saltenv=metalk8s-<current_cluster_version> pillar="{'orchestrate': {'node_name': <node_name>}}"
```

**Implementation proposal** (strongly recommended):

Create a bash script (like the bootstrap.sh one) that will:
- take a destination version as arguments
- execute the right state according to the downgrade procedure

This script should:
- only output few lines (like bootstrap script)
- have a verbose mode (like bootstrap script)
- have a log file (like bootstrap script)
- have a dry run mode to only check what will be done (maybe only running the `precheck` state)

Update the operation guide documentation

### PR DESCRIPTION
**Pull request by sayf-eddine-scality**
Fixes: #1403

**Component**: downgrade

**Context**: 
We need a script to automate downgrade procedure
**Summary**:
A script has been added to automate downgrade
**Acceptance criteria**: 

---

Closes: #1403

### CODE REVIEW

File: scripts/downgrade.sh [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      _usage
+      exit 1
+      ;;
+  esac
+done
+
+TMPFILES=$(mktemp -d)
+
+cat << EOF >> "${LOGFILE}"
+--- Downgrade script started on $(date -u -R) ---
```
Comment (TeddyAndrieux) [inline scripts/downgrade.sh]:
```suggestion
--- MetalK8s downgrade started on $(date -u -R) ---
```
To be consistent

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

File: scripts/downgrade.sh [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  $SALT salt-run state.sls metalk8s.orchestrate.deploy_node saltenv="$SALTENV" \
+    pillar="{'orchestrate': {'node_name': 'bootstrap'}}"
+
+}
+# Main
+_init
+run "Precheck downgrade" precheck_downgrade
+[ $DRY_RUN -eq 1 ] && exit 0
+run "Launch the downgrade" launch_downgrade
+run "Downgrade bootstrap" downgrade_bootstrap
```
Comment (TeddyAndrieux) [inline scripts/downgrade.sh]: To be consistent with upgrade and bootstrap we should add `ing` everywhere (Downgrading, Launching, ...) :) (btw I think some missing in iso manager)

File: scripts/downgrade.sh [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+die() {
+    echo 1>&2 "$@"
+    return 1
+}
+
+_set_env() {
+    if [ -z "$SALTENV" ]; then
+        SALTENV="metalk8s-$($SALT_CALL --out txt slsutil.renderer \
```
Comment (TeddyAndrieux) [inline scripts/downgrade.sh:138]: Hmm here you get the current bootstrap version but what we want it's the highest version installed and available but it's not really simple to get this one especially in a bash script :/ We should have an external pillar that will retrieve this. But maybe lets take the bootstrap version for this PR

File: scripts/downgrade.sh [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    metalk8s.scality.com/version \
+    node=bootstrap value="$DESTINATION_VERSION" \
+    kubeconfig=/etc/kubernetes/admin.conf
+  $SALT salt-run state.sls metalk8s.orchestrate.deploy_node saltenv="$SALTENV" \
+    pillar="{'orchestrate': {'node_name': 'bootstrap'}}"
+
+}
+# Main
+_init
+run "recheck" precheck_downgrade
```
Comment (TeddyAndrieux) [inline scripts/downgrade.sh]: "recheck" :stuck_out_tongue:

Comment (TeddyAndrieux) [reply scripts/downgrade.sh]: Also I prefer something like in @Ebaneck PR `Performing Pre-Downgrade checks` or `Performing Downgrade prechecks` as you want (both)

File: scripts/downgrade.sh (hunk 3 above)
Comment (sayf-eddine-scality) [reply scripts/downgrade.sh:138]: Yes this one should be rethought, lets create a ticket

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

File: scripts/downgrade.sh (hunk 3 above)
Comment (TeddyAndrieux) [reply scripts/downgrade.sh:138]: Btw we should also not downgrade/upgrade the bootstrap if it's already older/newer than the destination version. But not in these two PR I think

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

File: scripts/downgrade.sh (hunk 3 above)
Comment (sayf-eddine-scality) [reply scripts/downgrade.sh:138]: that should be in the precheck no ?

File: scripts/downgrade.sh [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    pillar="{'metalk8s': {'endpoints': $(salt-call \
+    --out txt pillar.get metalk8s:endpoints | cut -c 8-)}}" \
+    --retcode-passthrough
+  _check_salt_master
+  $SALT salt-run salt.cmd state.single metalk8s_kubernetes.node_label_present \
+    metalk8s.scality.com/version \
+    node=bootstrap value="$DESTINATION_VERSION" \
+    kubeconfig=/etc/kubernetes/admin.conf
+  $SALT salt-run state.sls metalk8s.orchestrate.deploy_node saltenv="$SALTENV" \
+    pillar="{'orchestrate': {'node_name': 'bootstrap'}}"
```
Comment (TeddyAndrieux) [inline scripts/downgrade.sh]: :cry: `bootstrap` :( 
```
[CRITICAL] Rendering SLS 'metalk8s-2.2.0-dev:metalk8s.orchestrate.deploy_node' failed: Jinja variable 'dict object' has no attribute u'bootstrap'
```
Should be the id of the bootstrap node

Comment (sayf-eddine-scality) [reply scripts/downgrade.sh]: oups, fixed

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

File: scripts/downgrade.sh (hunk 3 above)
Comment (TeddyAndrieux) [reply scripts/downgrade.sh:138]: That can't because we can downgrade/upgrade just we should skip the node already older/newer

**Comment by sayf-eddine-scality** [comment]
@bert-e approve

**Comment by sayf-eddine-scality** [comment]
Done berte

File: scripts/downgrade.sh [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    | awk '/^local\: /{ print $2 }')
+  [ -z "$bootstrap_id" ] && die "Cannot retrieve bootstrap id"
+  $SALT salt-run salt.cmd state.single metalk8s_kubernetes.node_label_present \
+    metalk8s.scality.com/version \
+    node="$bootstrap_id" value="$DESTINATION_VERSION" \
+    kubeconfig=/etc/kubernetes/admin.conf
+  $SALT salt-run state.sls metalk8s.orchestrate.deploy_node saltenv="$SALTENV" \
+    pillar="{'orchestrate': {'node_name': '$bootstrap_id'}}"
+
+}
```
Comment (TeddyAndrieux) [inline scripts/downgrade.sh:195]: minor: But remove the empty line before and add one after the bracket :)

**Review by TeddyAndrieux — APPROVED** [review]
LGTM

**Comment by TeddyAndrieux** [comment]
@bert-e reset

**Comment by sayf-eddine-scality** [comment]
Done berte

**Pull request merged**